internal/api/routes_heresphere.go

```diff
@@ -0,0 +1,1402 @@
+		r.Get("/", rs.HeresphereIndex)
+		r.Head("/", rs.HeresphereIndex)
+
+		r.Post("/auth", rs.HeresphereLoginToken)
+		r.Route("/{sceneId}", func(r chi.Router) {
+			r.Use(rs.HeresphereSceneCtx)
+
+			r.Post("/", rs.HeresphereVideoData)
+			r.Get("/", rs.HeresphereVideoData)
+
+			r.Post("/event", rs.HeresphereVideoEvent)
+		})
+	})
+
+	return r
+}
+
+func getVrTag() (varTag string, err error) {
+	// Find setting
+	varTag = config.GetInstance().GetUIVRTag()
+	if len(varTag) == 0 {
+		err = fmt.Errorf("zero length vr tag")
+	}
+	return
+}
+func getMinPlayPercent() (per int, err error) {
+	per = config.GetInstance().GetUIMinPlayPercent()
+	if per < 0 {
+		err = fmt.Errorf("unset minimum play percent")
+	}
+	return
+}
+
+/*
+ * This auxiliary function searches for the "favorite" tag
+ */
+func (rs heresphereRoutes) getVideoFavorite(r *http.Request, scene *models.Scene) bool {
+	tag_ids, err := rs.resolver.Scene().Tags(r.Context(), scene)
+	if err == nil {
+		favTag := config.GetInstance().GetHSPFavoriteTag()
+		for _, tag := range tag_ids {
+			if tag.ID == favTag {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+/*
+ * This is a video playback event
+ * Intended for server-sided script playback.
+ * But since we dont need that, we just use it for timestamps.
+ */
+func (rs heresphereRoutes) HeresphereVideoEvent(w http.ResponseWriter, r *http.Request) {
+	scn := r.Context().Value(heresphereKey).(*models.Scene)
+
+	// Decode event
+	var event HeresphereVideoEvent
+	err := json.NewDecoder(r.Body).Decode(&event)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Add playDuration
+	newTime := event.Time / 1000
+	newDuration := 0.0
+	if newTime > scn.ResumeTime {
+		newDuration += (newTime - scn.ResumeTime)
+	}
+
+	// Update PlayCount
+	if per, err := getMinPlayPercent(); err == nil {
+		// Above min playback percent
+		if file := scn.Files.Primary(); file != nil && newTime/file.Duration > float64(per)/100.0 {
+			// Create update set
+			ret := &scene.UpdateSet{
+				ID: scn.ID,
+			}
+			ret.Partial = models.NewScenePartial()
+
+			// Unless we track playback, we cant know if its a "played" or skip event
+			if scn.PlayCount == 0 {
+				ret.Partial.PlayCount.Set = true
+				ret.Partial.PlayCount.Value = 1
+			}
+
+			// Update scene
+			if err := txn.WithTxn(r.Context(), rs.txnManager, func(ctx context.Context) error {
+				_, err := ret.Update(ctx, rs.repository.Scene)
+				return err
+			}); err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+		}
+	}
+
+	// Write
+	if _, err := rs.resolver.Mutation().SceneSaveActivity(r.Context(), strconv.Itoa(scn.ID), &newTime, &newDuration); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+/*
+ * This endpoint is for letting the user update scene data
+ */
+func (rs heresphereRoutes) HeresphereVideoDataUpdate(w http.ResponseWriter, r *http.Request) error {
```

Comment:
This function is massive and needs to be split up.

Reply:
Done!